### PR TITLE
Align raster layers using QGIS functionality

### PR DESCRIPTION
**Pull request by wonder-sk**
The PR implements a new way of clipping and resampling of rasters within InaSAFE - see RFC #1998 for more details.

In order to test the new functionality one needs QGIS raster alignment tool - currently only available in my branch:
https://github.com/wonder-sk/QGIS/tree/raster-alignment-tool

If the QGIS raster alignment tool is not present, InaSAFE will use the original code path.

### CODE REVIEW

**Comment by wonder-sk** [comment]
The existing InaSAFE clipper does not seem to correctly clip at the border of the raster. Compare the clipped population layer when running flooding IF using jakarta 2007 flood, with full analysis extent...

Clipped population with existing InaSAFE clipper:
image

Clipped population with new QgsAlignRaster class:
image

As can be seen on the pictures, the existing clipper does not output the rightmost original pixels. Because of this, there are some differences in the output of impact functions - e.g. with the above mentioned flooding IF the total amount of people flooded > 1 meters is:
- 979K with existing InaSAFE clipper
- 1001K with new QgsAlignRaster

**Comment by akbargumbira** [comment]
Hmm, before doing the clipping we do some manipulation to the requested extent so that it doesn't clip some part of the pixels and fully include it as a pixel. Actually strange that the right most pixels seem to be clipped even with InaSAFE clipper.

**Comment by wonder-sk** [comment]
The origin, cell size and raster size is the same between the existing and new resampler/clipper, so the extent is adjusted correctly. Probably something is going wrong in gdalwarp.

**Comment by akbargumbira** [comment]
Right, I just looked at your code and you do also use the manipulated extent. So yeah, it must be in gdalwarp. LGTM for the logic, but probably Tim will buzz you with the coding standards :)

File: safe/utilities/clipper.py [hunk 1]
```diff
@@ -676,3 +677,107 @@ def adjust_clip_extent(clip_extent, cell_size, layer_extent):
...
+           QgsAlignRaster.Item(exposure_layer.source(), exposure_output)]
+    # TODO: what keywords to consider for rescaling?
+    if exposure_keywords.get('exposure_unit') == 'count':
+        lst[1].rescaleValues = True
+    align.setRasters(lst)
+
+    # Find out which layer has finer grid and use it as the reference.
+    # This will setup destination CRS, cell size and grid origin
+    # TODO: need to respect 'allow_resampling' keyword?
+    if exposure_keywords.get('allow_resampling', 'true').lower() == 'true':
```
Comment (akbargumbira) [inline safe/utilities/clipper.py]: probably would be better to have a more generic method of this (i.e having one parameter to which layer it should be resampled regarding the grid size) and passing the preference for that would be the job of analysis class (and moving the keywords 'allow_resampling' to that class). With that, this function would be also useful for testing native IF's as we need to clip the layer first. To do that now, I made a clipping function for testing purpose which a bit dodgy

Comment (akbargumbira) [inline safe/utilities/clipper.py]: See this for what I meant https://github.com/AIFDR/inasafe/blob/develop/safe/test/utilities.py#L1181

**Comment by timlinux** [comment]
@wonder-sk Actually I always find it strange that QGIS renders half pixel widths like in both your screenies on the left and right edges. It doesn't make any sense to me why it does it since I thought it was standard in raster data definition to work only in whole pixel units.

With regards to small differences in cells being returned, I think that is ok, especially if the new clipper produces something with better fidelity to the intended clip area. We will need to put logic in our tests to have different expected results though based on whether ``QgsAlignRaster`` was used or not....which is probably going to be a little tedious but I can't see another way past it.

**Comment by wonder-sk** [comment]
@timlinux There is nothing wrong with QGIS rendering here: the layers actually have 638 x 649 pixels, so what looks like single pixels is actually composed from many small pixels... no pixels were cut in half or harmed in any other way.

**Comment by timlinux** [comment]
@wonder-sk thanks for not harming the pixels :-)

Can you please update this PR to merge against experimental branch.

**Comment by wonder-sk** [comment]
@akbargumbira Made qgis_align_rasters() more generic in 15115e1

**Comment by wonder-sk** [comment]
Merged to experimental branch.

**Pull request closed without merging**